validator/client/src/test/java/tech/pegasys/teku/validator/client/ValidatorRegistratorTest.java

```diff
@@ -151,19 +158,98 @@ void registersValidators_onBeginningOfEpoch() {
         captureValidatorApiChannelCalls(2);
 
     registrationCalls.forEach(
-        registrationCall ->
-            verifyRegistrations(
-                registrationCall,
-                List.of(
-                    validator1.getPublicKey(),
-                    validator2.getPublicKey(),
-                    validator3.getPublicKey())));
+        registrationCall -> verifyRegistrations(registrationCall, allValidatorsPublicKeys));
 
     // signer will be called in total 3 times, since from the 2nd run the signed registrations will
     // be cached
     verify(signer, times(3)).signValidatorRegistration(any(), any());
   }
 
+  @TestTemplate
+  void cleanupsCache_ifValidatorIsNoLongerActive() {
+    // GIVEN
+    when(ownedValidators.getActiveValidators())
+        .thenReturn(List.of(validator1, validator2, validator3));
+
+    // WHEN first invoked, registrations will be cached
+    validatorRegistrator.onSlot(UInt64.ZERO);
+
+    // THEN registrations should be cached
+    assertThat(validatorRegistrator.getNumberOfCachedRegistrations()).isEqualTo(3);
+
+    // GIVEN validator1 is not active anymore
+    when(ownedValidators.getActiveValidators()).thenReturn(List.of(validator2, validator3));
+
+    // WHEN invoked again
+    validatorRegistrator.onSlot(UInt64.valueOf(slotsPerEpoch));
+
+    // THEN cache should be cleaned up
+    assertThat(validatorRegistrator.getNumberOfCachedRegistrations()).isEqualTo(2);
+  }
+
+  @TestTemplate
+  void doesNotUseCache_ifRegistrationsNeedUpdating() {
```

Comment:
this is quite large for a unit test, should bits of setup or something be refactored so that it's easier to read the actual test?